torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -1373,72 +1373,76 @@ void InitXlaModuleBindings(py::module m) {
                 weight_decay, eps, amsgrad, maximize, use_adamw);
           }
         });
-  m.def("_xla_mark_sharding",
-        [](const at::Tensor& input, const py::list& tile_assignment,
-           const py::list& group_assignment, const py::list& replication_groups,
-           int sharding_type) {
-          TORCH_LAZY_COUNTER("XlaMarkSharding", 1);
-          XLA_CHECK(UseVirtualDevice()) << "Please set `XLA_USE_SPMD=1`";
-          XLATensorPtr xtensor = bridge::GetXlaTensor(input);
-          xla::OpSharding sharding = ShardingUtil::CreateOpSharding(
-              tile_assignment, group_assignment, replication_groups,
-              ShardingUtil::ShardingType(sharding_type));
-          auto new_sharding_spec = std::make_shared<XLATensor::ShardingSpec>(
-              sharding,
-              MakeShapeWithDeviceLayout(
-                  xtensor->shape(),
-                  static_cast<XlaDeviceType>(xtensor->GetDevice().type())));
-
-          // For IR values, we directly attach the sharding spec to the xtensor.
-          if (xtensor->CurrentIrValue()) {
-            // TODO(alanwaketan): Do we want to check if there is any existing
-            // sharding spec? It seems okay to directly overwrite it.
-            xtensor->SetShardingSpec(*new_sharding_spec);
-            return;
-          }
+  m.def("_xla_mark_sharding", [](const at::Tensor& input,
+                                 const py::list& tile_assignment,
+                                 const py::list& group_assignment,
+                                 const py::list& replication_groups,
+                                 int sharding_type) {
+    TORCH_LAZY_COUNTER("XlaMarkSharding", 1);
+    XLA_CHECK(UseVirtualDevice()) << "Please set `XLA_USE_SPMD=1`";
+    XLATensorPtr xtensor = bridge::GetXlaTensor(input);
+    xla::OpSharding sharding = ShardingUtil::CreateOpSharding(
+        tile_assignment, group_assignment, replication_groups,
+        ShardingUtil::ShardingType(sharding_type));
+    auto new_sharding_spec = std::make_shared<XLATensor::ShardingSpec>(
+        sharding, MakeShapeWithDeviceLayout(
+                      xtensor->shape(),
+                      static_cast<XlaDeviceType>(xtensor->GetDevice().type())));
+
+    // For Non DeviceData IR values, we directly attach the sharding spec
+    // to the xtensor.
+    const DeviceData* device_data_node = nullptr;
```

Comment:
I don't know why formater went crazy when I add this line, I can't seems to overwrite it. the real change here is to check whether the tensor is a Device Data, if yes then fallback to the old path to use `XLAGraphExecutor::Get()->GetTensors` and move the tensor back to CPU. Everything else is formatting change.

Reply:
Lol, thanks!